src/olympia/addons/serializers.py

```diff
@@ -486,7 +487,10 @@ def validate_upload(self, value):
         if not own_upload or not value.valid or value.validation_timeout:
             raise exceptions.ValidationError(gettext('Upload is not valid.'))
         # Parse the file to get and validate package data with the addon.
-        self.parsed_data = parse_addon(value, addon=self.addon, user=request.user)
+        try:
+            self.parsed_data = parse_addon(value, addon=self.addon, user=request.user)
+        except DuplicateAddonID as exc:
+            raise Conflict({self.field_name: exc.messages})
```

Comment:
I guess you need to manually wrap the `exc.message` in a dict here (exceptions from a `validate_xxx` function should already get wrapped as `{xxx: exception}`) because an `APIException` subclass is being raised rather than a `ValidationError`?

Reply:
Yes. That is annoying, but there is no method to call to do that, the `ValidationError` conversion happens in DRF `is_valid()` code (and there is something else for authentication errors or 500 exceptions).

I have raised a non-dict, but then it wouldn't have been consistent with the way errors are typically returned.